{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# DataPath Example 4\n",
    "\n",
    "This notebook covers somewhat more advanced examples for using `DataPath`s. It assumes that you understand \n",
    "the concepts presented in the previous example notebooks.\n",
    "\n",
    "You should also read the ERMrest documentation and the derivapy wiki. There are more advanced concepts in this notebook that are demonstrated but not fully (re)explained here, as the concepts are explained in other documentation.\n",
    "\n",
    "## Exampe Data Model\n",
    "The examples require that you understand a little bit about the example catalog data model, which is based on the FaceBase project.\n",
    "\n",
    "### Key tables\n",
    "- `'dataset'` : represents a unit of data usually a `'study'` or `'experiment'`\n",
    "- `'sample'` : a biosample\n",
    "- `'assay'` : a bioassay (typically RNA-seq or ChIP-seq assays)\n",
    "\n",
    "### Relationships\n",
    "- `dataset <- sample`: A dataset may have one to many samples. I.e., there \n",
    "  is a foreign key reference from sample to dataset.\n",
    "- `sample <- assay`: A sample may have one to many assays. I.e., there is a\n",
    "  foreign key reference from assay to sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Import deriva modules\n",
    "from deriva_common import ErmrestCatalog, get_credential"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Connect with the deriva catalog\n",
    "protocol = 'https'\n",
    "hostname = 'www.facebase.org'\n",
    "catalog_number = 1\n",
    "credential = None\n",
    "# If you need to authenticate, use Deriva Auth agent and get the credential\n",
    "# credential = get_credential(hostname)\n",
    "catalog = ErmrestCatalog(protocol, hostname, catalog_number, credential)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Get the path builder interface for this catalog\n",
    "pb = catalog.getPathBuilder()\n",
    "\n",
    "# Get some local variable handles to tables for convenience\n",
    "dataset = pb.isa.dataset\n",
    "sample = pb.isa.sample\n",
    "assay = pb.isa.assay"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Implicit DataPaths\n",
    "**Proceed with caution**\n",
    "\n",
    "For compactness, `Table` objects (and `TableAlias` objects) provide `DataPath`-like methods. E.g., `link(...)`, `filter(...)`, and `entities(...)`, which will implicitly create `DataPath`s rooted at the table and return the newly created path. These operations `return` the new `DataPath` rather than mutating the `Table` (or `TableAlias`) objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities = dataset.filter(dataset.status > 2).entities()\n",
    "len(entities)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DataPath-like methods\n",
    "The `DataPath`-like methods on `Table`s are essentially \"wrapper\" functions over the implicitly generated `DataPath` rooted at the `Table` instance. The wrappers include:\n",
    "- `link(...)` generates a path, links the related table, and returns the generated path.\n",
    "- `filter(...)` generate a path, adds a filter, and returns the generated path.\n",
    "- `entities(...)` generate a path, (optionally) selects columns, and returns an entity set."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Attribute Examples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: selecting all columns of a table instance\n",
    "Passing a table (or table instance) object to the `entities(...)` method will project all (i.e., `*`) of its attributes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "171\n",
      "https://www.facebase.org/ermrest/catalog/1/entity/D:=isa:dataset/sample:=isa:sample/assay:=isa:assay\n"
     ]
    }
   ],
   "source": [
    "path = dataset.alias('D').path\n",
    "path.link(sample).link(assay)\n",
    "entities = path.entities()\n",
    "print(len(entities))\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/attribute/D:=isa:dataset/sample:=isa:sample/assay:=isa:assay/D:*\n"
     ]
    }
   ],
   "source": [
    "entities = path.entities(path.D)\n",
    "print(entities.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is important to remember that the entities method returns a entity set based on the entity type of the last elmenent of the path. In this example that means the number of results will be determined by the number of unique rows in the assay table instance in the path created above, as the last link method used the assay table.  \n",
    "\n",
    "In this example, we are only using attributes from dataset.  What this means is that even though the result of an entities method is always a entity set, since we are not including the attributes from assay in our result, it is possible that we may have duplicate rows, if the same assay is used in more then one dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: selecting from multiple table instances\n",
    "More than one table instance may be selected in this manner and it can be mixed and matched with columns from other tables instances. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "171\n",
      "https://www.facebase.org/ermrest/catalog/1/attribute/D:=isa:dataset/sample:=isa:sample/assay:=isa:assay/D:*,assay:fragmentation_method,sample:*,assay:isolation_protocol\n"
     ]
    }
   ],
   "source": [
    "entities = path.entities(path.D,\n",
    "                         path.assay.fragmentation_method,\n",
    "                         path.sample,\n",
    "                         path.assay.isolation_protocol)\n",
    "print (len(entities))\n",
    "print(entities.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " If you want to base the results on a different entity, you can introduce a table instance alias into the end of the path, before calling the entities function.  In this case, even though we are asking for the same attributes, we are getting the set of datasets, not the set of assays.  Also, since we are including the attributes from dataset in our query, we know that we will not be seeing any duplicate rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n",
      "https://www.facebase.org/ermrest/catalog/1/attribute/D:=isa:dataset/sample:=isa:sample/assay:=isa:assay/$D/D:*,assay:fragmentation_method,sample:*,assay:isolation_protocol\n"
     ]
    }
   ],
   "source": [
    "entities = path.D.entities(path.D,\n",
    "                           path.assay.fragmentation_method,\n",
    "                           path.sample,\n",
    "                           path.assay.isolation_protocol)\n",
    "print (len(entities))\n",
    "print(entities.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering Examples\n",
    "\n",
    "### Example: filters with logical operators\n",
    "This example shows how to combine two comparisons with a conjuncting (i.e., `and` operator). Because Python's logical-and (`and`) keyword cannot be overloaded, we instead overload the bitwise-and (`&`) operator. This approach has become customary among many similar data access libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/sample:=isa:sample/assay:=isa:assay/(sample_type=ChIP-seq)&(selection=H3K27AC)\n"
     ]
    }
   ],
   "source": [
    "path = dataset.link(sample).link(assay).filter(\n",
    "    ((assay.sample_type == 'ChIP-seq') & (assay.selection == 'H3K27AC')))\n",
    "\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alignment_id</th>\n",
       "      <th>cell_count</th>\n",
       "      <th>dataset</th>\n",
       "      <th>fragmentation_method</th>\n",
       "      <th>id</th>\n",
       "      <th>isolation_protocol</th>\n",
       "      <th>library_id</th>\n",
       "      <th>markers</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>pretreatment</th>\n",
       "      <th>...</th>\n",
       "      <th>reagent_batch_number</th>\n",
       "      <th>reagent_catalog_number</th>\n",
       "      <th>reagent_source</th>\n",
       "      <th>replicate</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_composition</th>\n",
       "      <th>sample_purification</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>selection</th>\n",
       "      <th>tracks_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>45</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>65</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007.0</td>\n",
       "      <td>39133.0</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>medial nasal process</td>\n",
       "      <td>excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>52</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>72</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007.0</td>\n",
       "      <td>39133.0</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>mandibular process</td>\n",
       "      <td>excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>54</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>14</td>\n",
       "      <td></td>\n",
       "      <td>74</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007.0</td>\n",
       "      <td>39133.0</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>latero nasal process</td>\n",
       "      <td>excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>59</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>19</td>\n",
       "      <td></td>\n",
       "      <td>79</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007.0</td>\n",
       "      <td>39133.0</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>maxillary process</td>\n",
       "      <td>excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>61</td>\n",
       "      <td>NA</td>\n",
       "      <td>14130</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>21</td>\n",
       "      <td>None</td>\n",
       "      <td>81</td>\n",
       "      <td>Histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007</td>\n",
       "      <td>39133</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>1088</td>\n",
       "      <td>face</td>\n",
       "      <td>Excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>63</td>\n",
       "      <td>NA</td>\n",
       "      <td>14130</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>26</td>\n",
       "      <td>None</td>\n",
       "      <td>86</td>\n",
       "      <td>Histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007</td>\n",
       "      <td>39133</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>1089</td>\n",
       "      <td>face</td>\n",
       "      <td>Excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   alignment_id cell_count  dataset fragmentation_method  id  \\\n",
       "0            45         NA    14068           Sonication   5   \n",
       "1            52         NA    14068           Sonication  12   \n",
       "2            54         NA    14068           Sonication  14   \n",
       "3            59         NA    14068           Sonication  19   \n",
       "4            61         NA    14130           Sonication  21   \n",
       "5            63         NA    14130           Sonication  26   \n",
       "\n",
       "  isolation_protocol  library_id    markers molecule_type        pretreatment  \\\n",
       "0                             65  histology    genomicDNA  Formaldehyde fixed   \n",
       "1                             72  histology    genomicDNA  Formaldehyde fixed   \n",
       "2                             74  histology    genomicDNA  Formaldehyde fixed   \n",
       "3                             79  histology    genomicDNA  Formaldehyde fixed   \n",
       "4               None          81  Histology    genomicDNA  Formaldehyde fixed   \n",
       "5               None          86  Histology    genomicDNA  Formaldehyde fixed   \n",
       "\n",
       "     ...    reagent_batch_number reagent_catalog_number reagent_source  \\\n",
       "0    ...               1613007.0                39133.0   Active Motif   \n",
       "1    ...               1613007.0                39133.0   Active Motif   \n",
       "2    ...               1613007.0                39133.0   Active Motif   \n",
       "3    ...               1613007.0                39133.0   Active Motif   \n",
       "4    ...                 1613007                  39133   Active Motif   \n",
       "5    ...                 1613007                  39133   Active Motif   \n",
       "\n",
       "  replicate sample    sample_composition  sample_purification sample_type  \\\n",
       "0         1      1  medial nasal process             excision    ChIP-seq   \n",
       "1         1      4    mandibular process             excision    ChIP-seq   \n",
       "2         1      3  latero nasal process             excision    ChIP-seq   \n",
       "3         1      2     maxillary process             excision    ChIP-seq   \n",
       "4         1   1088                  face             Excision    ChIP-seq   \n",
       "5         1   1089                  face             Excision    ChIP-seq   \n",
       "\n",
       "  selection tracks_id  \n",
       "0   H3K27AC        25  \n",
       "1   H3K27AC        32  \n",
       "2   H3K27AC        34  \n",
       "3   H3K27AC        39  \n",
       "4   H3K27AC        41  \n",
       "5   H3K27AC        42  \n",
       "\n",
       "[6 rows x 22 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path.entities().dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: combine conjunction and disjunctions in filters\n",
    "Similar to the prior example, the filters allow combining of conjunctive and disjunctive operators. Like the bitwise-and operator, we also overload the bitwise-or (` | `) operator because the logical-or (`or`) operatar cannot be overloaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/sample:=isa:sample/assay:=isa:assay/((sample_type=ChIP-seq)&(selection=H3K27AC));((sample_type=Input)&(selection=genomicDNA))\n"
     ]
    }
   ],
   "source": [
    "path = dataset.link(sample).link(assay).filter(\n",
    "    ((assay.sample_type == 'ChIP-seq') & (assay.selection == 'H3K27AC')) |\n",
    "    ((assay.sample_type == 'Input') & (assay.selection == 'genomicDNA')))\n",
    "\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alignment_id</th>\n",
       "      <th>cell_count</th>\n",
       "      <th>dataset</th>\n",
       "      <th>fragmentation_method</th>\n",
       "      <th>id</th>\n",
       "      <th>isolation_protocol</th>\n",
       "      <th>library_id</th>\n",
       "      <th>markers</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>pretreatment</th>\n",
       "      <th>...</th>\n",
       "      <th>reagent_batch_number</th>\n",
       "      <th>reagent_catalog_number</th>\n",
       "      <th>reagent_source</th>\n",
       "      <th>replicate</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_composition</th>\n",
       "      <th>sample_purification</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>selection</th>\n",
       "      <th>tracks_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>42</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>62</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>medial nasal process</td>\n",
       "      <td>excision</td>\n",
       "      <td>Input</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>45</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td>65</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007.0</td>\n",
       "      <td>39133.0</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>medial nasal process</td>\n",
       "      <td>excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>47</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>7</td>\n",
       "      <td></td>\n",
       "      <td>67</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>latero nasal process</td>\n",
       "      <td>excision</td>\n",
       "      <td>Input</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>49</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>9</td>\n",
       "      <td></td>\n",
       "      <td>69</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>mandibular process</td>\n",
       "      <td>excision</td>\n",
       "      <td>Input</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>52</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>12</td>\n",
       "      <td></td>\n",
       "      <td>72</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007.0</td>\n",
       "      <td>39133.0</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>mandibular process</td>\n",
       "      <td>excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>54</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>14</td>\n",
       "      <td></td>\n",
       "      <td>74</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007.0</td>\n",
       "      <td>39133.0</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>latero nasal process</td>\n",
       "      <td>excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>56</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>16</td>\n",
       "      <td></td>\n",
       "      <td>76</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>maxillary process</td>\n",
       "      <td>excision</td>\n",
       "      <td>Input</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>59</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>19</td>\n",
       "      <td></td>\n",
       "      <td>79</td>\n",
       "      <td>histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007.0</td>\n",
       "      <td>39133.0</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>maxillary process</td>\n",
       "      <td>excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>39.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>61</td>\n",
       "      <td>NA</td>\n",
       "      <td>14130</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>21</td>\n",
       "      <td>None</td>\n",
       "      <td>81</td>\n",
       "      <td>Histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007</td>\n",
       "      <td>39133</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>1088</td>\n",
       "      <td>face</td>\n",
       "      <td>Excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>61</td>\n",
       "      <td>NA</td>\n",
       "      <td>14130</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>24</td>\n",
       "      <td>None</td>\n",
       "      <td>84</td>\n",
       "      <td>Histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>1088</td>\n",
       "      <td>face</td>\n",
       "      <td>Excision</td>\n",
       "      <td>Input</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>63</td>\n",
       "      <td>NA</td>\n",
       "      <td>14130</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>26</td>\n",
       "      <td>None</td>\n",
       "      <td>86</td>\n",
       "      <td>Histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>1613007</td>\n",
       "      <td>39133</td>\n",
       "      <td>Active Motif</td>\n",
       "      <td>1</td>\n",
       "      <td>1089</td>\n",
       "      <td>face</td>\n",
       "      <td>Excision</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>H3K27AC</td>\n",
       "      <td>42.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>63</td>\n",
       "      <td>NA</td>\n",
       "      <td>14130</td>\n",
       "      <td>Sonication</td>\n",
       "      <td>29</td>\n",
       "      <td>None</td>\n",
       "      <td>89</td>\n",
       "      <td>Histology</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>Formaldehyde fixed</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>1089</td>\n",
       "      <td>face</td>\n",
       "      <td>Excision</td>\n",
       "      <td>Input</td>\n",
       "      <td>genomicDNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    alignment_id cell_count  dataset fragmentation_method  id  \\\n",
       "0             42         NA    14068           Sonication   2   \n",
       "1             45         NA    14068           Sonication   5   \n",
       "2             47         NA    14068           Sonication   7   \n",
       "3             49         NA    14068           Sonication   9   \n",
       "4             52         NA    14068           Sonication  12   \n",
       "5             54         NA    14068           Sonication  14   \n",
       "6             56         NA    14068           Sonication  16   \n",
       "7             59         NA    14068           Sonication  19   \n",
       "8             61         NA    14130           Sonication  21   \n",
       "9             61         NA    14130           Sonication  24   \n",
       "10            63         NA    14130           Sonication  26   \n",
       "11            63         NA    14130           Sonication  29   \n",
       "\n",
       "   isolation_protocol  library_id    markers molecule_type  \\\n",
       "0                              62  histology    genomicDNA   \n",
       "1                              65  histology    genomicDNA   \n",
       "2                              67  histology    genomicDNA   \n",
       "3                              69  histology    genomicDNA   \n",
       "4                              72  histology    genomicDNA   \n",
       "5                              74  histology    genomicDNA   \n",
       "6                              76  histology    genomicDNA   \n",
       "7                              79  histology    genomicDNA   \n",
       "8                None          81  Histology    genomicDNA   \n",
       "9                None          84  Histology    genomicDNA   \n",
       "10               None          86  Histology    genomicDNA   \n",
       "11               None          89  Histology    genomicDNA   \n",
       "\n",
       "          pretreatment    ...    reagent_batch_number reagent_catalog_number  \\\n",
       "0   Formaldehyde fixed    ...                      NA                     NA   \n",
       "1   Formaldehyde fixed    ...               1613007.0                39133.0   \n",
       "2   Formaldehyde fixed    ...                      NA                     NA   \n",
       "3   Formaldehyde fixed    ...                      NA                     NA   \n",
       "4   Formaldehyde fixed    ...               1613007.0                39133.0   \n",
       "5   Formaldehyde fixed    ...               1613007.0                39133.0   \n",
       "6   Formaldehyde fixed    ...                      NA                     NA   \n",
       "7   Formaldehyde fixed    ...               1613007.0                39133.0   \n",
       "8   Formaldehyde fixed    ...                 1613007                  39133   \n",
       "9   Formaldehyde fixed    ...                      NA                     NA   \n",
       "10  Formaldehyde fixed    ...                 1613007                  39133   \n",
       "11  Formaldehyde fixed    ...                      NA                     NA   \n",
       "\n",
       "   reagent_source replicate sample    sample_composition  sample_purification  \\\n",
       "0              NA         4      1  medial nasal process             excision   \n",
       "1    Active Motif         1      1  medial nasal process             excision   \n",
       "2              NA         4      3  latero nasal process             excision   \n",
       "3              NA         4      4    mandibular process             excision   \n",
       "4    Active Motif         1      4    mandibular process             excision   \n",
       "5    Active Motif         1      3  latero nasal process             excision   \n",
       "6              NA         4      2     maxillary process             excision   \n",
       "7    Active Motif         1      2     maxillary process             excision   \n",
       "8    Active Motif         1   1088                  face             Excision   \n",
       "9              NA         4   1088                  face             Excision   \n",
       "10   Active Motif         1   1089                  face             Excision   \n",
       "11             NA         4   1089                  face             Excision   \n",
       "\n",
       "   sample_type   selection tracks_id  \n",
       "0        Input  genomicDNA      22.0  \n",
       "1     ChIP-seq     H3K27AC      25.0  \n",
       "2        Input  genomicDNA      27.0  \n",
       "3        Input  genomicDNA      29.0  \n",
       "4     ChIP-seq     H3K27AC      32.0  \n",
       "5     ChIP-seq     H3K27AC      34.0  \n",
       "6        Input  genomicDNA      36.0  \n",
       "7     ChIP-seq     H3K27AC      39.0  \n",
       "8     ChIP-seq     H3K27AC      41.0  \n",
       "9        Input  genomicDNA       NaN  \n",
       "10    ChIP-seq     H3K27AC      42.0  \n",
       "11       Input  genomicDNA       NaN  \n",
       "\n",
       "[12 rows x 22 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path.entities().dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: filtering at different stages of the path\n",
    "Filtering a path does not have to be done at the end of a path. In fact, the initial intention of the ERMrest URI was to mimick \"RESTful\" semantics where a RESTful \"resource\" is identified, then filtered, then a \"sub-resource\" is identified, and then filtered, and so on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/release_date::geq::2017-01-01/sample:=isa:sample/species=2/assay:=isa:assay/selection=totalRNA\n"
     ]
    }
   ],
   "source": [
    "path = dataset.filter(dataset.release_date >= '2017-01-01') \\\n",
    "    .link(sample).filter(sample.species == 2) \\\n",
    "    .link(assay).filter(assay.selection == 'totalRNA')\n",
    "    \n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alignment_id</th>\n",
       "      <th>cell_count</th>\n",
       "      <th>dataset</th>\n",
       "      <th>fragmentation_method</th>\n",
       "      <th>id</th>\n",
       "      <th>isolation_protocol</th>\n",
       "      <th>library_id</th>\n",
       "      <th>markers</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>pretreatment</th>\n",
       "      <th>...</th>\n",
       "      <th>reagent_batch_number</th>\n",
       "      <th>reagent_catalog_number</th>\n",
       "      <th>reagent_source</th>\n",
       "      <th>replicate</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_composition</th>\n",
       "      <th>sample_purification</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>selection</th>\n",
       "      <th>tracks_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>41</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Fragmentation Buffer from Illumina</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>61</td>\n",
       "      <td>histology</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Trizol</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>15032619.0</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>medial nasal process</td>\n",
       "      <td>excision</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>totalRNA</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>46</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Fragmentation Buffer from Illumina</td>\n",
       "      <td>6</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>histology</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Trizol</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>15032619.0</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>latero nasal process</td>\n",
       "      <td>excision</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>totalRNA</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>55</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Fragmentation Buffer from Illumina</td>\n",
       "      <td>15</td>\n",
       "      <td></td>\n",
       "      <td>75</td>\n",
       "      <td>histology</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Trizol</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>15032619.0</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>maxillary process</td>\n",
       "      <td>excision</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>totalRNA</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>60</td>\n",
       "      <td>NA</td>\n",
       "      <td>14068</td>\n",
       "      <td>Fragmentation Buffer from Illumina</td>\n",
       "      <td>20</td>\n",
       "      <td></td>\n",
       "      <td>80</td>\n",
       "      <td>histology</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Trizol</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>15032619.0</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>mandibular process</td>\n",
       "      <td>excision</td>\n",
       "      <td>RNA-seq</td>\n",
       "      <td>totalRNA</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   alignment_id cell_count  dataset                fragmentation_method  id  \\\n",
       "0            41         NA    14068  Fragmentation Buffer from Illumina   1   \n",
       "1            46         NA    14068  Fragmentation Buffer from Illumina   6   \n",
       "2            55         NA    14068  Fragmentation Buffer from Illumina  15   \n",
       "3            60         NA    14068  Fragmentation Buffer from Illumina  20   \n",
       "\n",
       "  isolation_protocol  library_id    markers molecule_type pretreatment  \\\n",
       "0                             61  histology          mRNA       Trizol   \n",
       "1                             66  histology          mRNA       Trizol   \n",
       "2                             75  histology          mRNA       Trizol   \n",
       "3                             80  histology          mRNA       Trizol   \n",
       "\n",
       "     ...    reagent_batch_number reagent_catalog_number reagent_source  \\\n",
       "0    ...                                     15032619.0       Illumina   \n",
       "1    ...                                     15032619.0       Illumina   \n",
       "2    ...                                     15032619.0       Illumina   \n",
       "3    ...                                     15032619.0       Illumina   \n",
       "\n",
       "  replicate sample    sample_composition  sample_purification sample_type  \\\n",
       "0         5      1  medial nasal process             excision     RNA-seq   \n",
       "1         5      3  latero nasal process             excision     RNA-seq   \n",
       "2         5      2     maxillary process             excision     RNA-seq   \n",
       "3         5      4    mandibular process             excision     RNA-seq   \n",
       "\n",
       "  selection tracks_id  \n",
       "0  totalRNA        21  \n",
       "1  totalRNA        26  \n",
       "2  totalRNA        35  \n",
       "3  totalRNA        40  \n",
       "\n",
       "[4 rows x 22 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path.entities().dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Linking Examples\n",
    "\n",
    "### Example: explicit column links\n",
    "Up until now, the examples have shown how to link entities via _implicit_ join predicates. That is, we knew there existed a foriegn key reference constraint between foreign keys of one entity and keys of another entity. We needed only to ask ERMrest to link the entities in order to get the linked set.\n",
    "\n",
    "The problem with implicit links is that it become _ambiguous_ if there are more than one foreign key reference between tables. To support these situations, ERMrest and the `DataPath`'s `link(...)` method can specify the columns to use for the link condition, explicitly.\n",
    "\n",
    "The structure of the `on` clause is:\n",
    "- an equality comparison operation where\n",
    "- the _left_ operand is a column of the _left_ table instance which is also the path _context_ before the link method is called, and\n",
    "- the _right_ operand is a column of the _right_ table instance which is the table _to be linked_ to the path."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/sample:=(id)=(isa:sample:dataset)\n"
     ]
    }
   ],
   "source": [
    "path = dataset.link(sample, on=(dataset.id==sample.dataset))\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**IMPORTANT** Not all tables are related by foreign key references. ERMrest does not allow arbitrary relational joins. Tables must be related by a foreign key reference in order to link them in a data path."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_keywords</th>\n",
       "      <th>anatomy</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>dataset</th>\n",
       "      <th>gender</th>\n",
       "      <th>gene</th>\n",
       "      <th>genotype</th>\n",
       "      <th>id</th>\n",
       "      <th>litter</th>\n",
       "      <th>local_identifier</th>\n",
       "      <th>mutation</th>\n",
       "      <th>origin</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>replicate</th>\n",
       "      <th>species</th>\n",
       "      <th>specimen</th>\n",
       "      <th>stage</th>\n",
       "      <th>strain</th>\n",
       "      <th>theiler_stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mus musculus wild type CD1 E16.5 Medial nasal ...</td>\n",
       "      <td>213</td>\n",
       "      <td>None</td>\n",
       "      <td>14068</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>E11.5_MNP</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>4</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mus musculus wild type CD1 E16.5 Maxillary pro...</td>\n",
       "      <td>207</td>\n",
       "      <td>None</td>\n",
       "      <td>14068</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>E11.5_MX</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>4</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mus musculus wild type CD1 E16.5 Lateral nasal...</td>\n",
       "      <td>167</td>\n",
       "      <td>None</td>\n",
       "      <td>14068</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>18</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "      <td>E11.5_LNP</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td>66</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>4</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           _keywords  anatomy collection_date  \\\n",
       "0  Mus musculus wild type CD1 E16.5 Medial nasal ...      213            None   \n",
       "1  Mus musculus wild type CD1 E16.5 Maxillary pro...      207            None   \n",
       "2  Mus musculus wild type CD1 E16.5 Lateral nasal...      167            None   \n",
       "\n",
       "   dataset gender  gene  genotype  id litter local_identifier mutation origin  \\\n",
       "0    14068   None  None        18   1               E11.5_MNP     None          \n",
       "1    14068   None  None        18   2                E11.5_MX     None          \n",
       "2    14068   None  None        18   3               E11.5_LNP     None          \n",
       "\n",
       "   phenotype replicate  species  specimen  stage  strain  theiler_stage  \n",
       "0         66      None        2        10     30       4             15  \n",
       "1         66      None        2        10     30       4             15  \n",
       "2         66      None        2        10     30       4             15  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path.entities().fetch(limit=3).dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: explicit column links combined with table aliasing\n",
    "As usual, table instances are generated automatically unless we provide a table alias."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/dataset:=isa:dataset/S:=(id)=(isa:sample:dataset)\n"
     ]
    }
   ],
   "source": [
    "path = dataset.link(sample.alias('S'), on=(dataset.id==sample.dataset))\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that we cannot use the alias right away in the `on` clause because it was not _bound_ to the path until _after_ the `link(...)` operation was performed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: links with \"outer join\" semantics\n",
    "Up until now, the examples have shown \"`link`s\" with _inner join_ semantics. _Outer join_ semantics can be expressed as part of explicit column links, and _only_ when using explicit column links.\n",
    "\n",
    "The `link(...)` method accepts a \"`join_type`\" parameter, i.e., \"`.link(... join_type=TYPE)`\", where _TYPE_ may be `'left'`, `'right'`, `'full'`, and defaults to `''` which indicates inner join type.\n",
    "\n",
    "By '`right`' outer joining in the link from `'sample'` to `'dataset`', the following path gives us a reference to `'dataset'` entities that _may or may not_ have any samples with assays."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/attribute/assay:=isa:assay/S:=(sample)=(isa:sample:id)/D:=right(dataset)=(isa:dataset:id)/D:accession,D:title,S:species,S:stage\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "712"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = assay.link(sample.alias('S'), on=assay.sample==sample.id) \\\n",
    "            .link(dataset.alias('D'), sample.dataset==dataset.id, join_type='right')\n",
    "\n",
    "# path instances were not bound until the above line was executed\n",
    "\n",
    "entities = path.entities(path.D.accession, path.D.title, path.S.species, path.S.stage)\n",
    "\n",
    "print(entities.uri)\n",
    "len(entities)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see above that we have a full set of datasets _whether or not_ they have any samples with assays. For further evidence, we can convert to a DataFrame and look at a slice of its entries. Note that the sample's 'species' and 'stage' attributes do not exist for some entities (i.e., `NaN`) because those entities did not exist for the join condition."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>species</th>\n",
       "      <th>stage</th>\n",
       "      <th>title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>700</th>\n",
       "      <td>FB00000886</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skull vault formation in individual WT AB Fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>FB00000887</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skull vault formation in individual WT AB Fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>FB00000888</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skull vault formation in individual WT AB Fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>703</th>\n",
       "      <td>FB00000889</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skull vault formation in individual WT AB Fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>FB00000890</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skull vault formation in individual WT AB Fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>FB00000807.2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>ChIP-seq of multiple histone marks and RNA-seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>FB00000867</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Temporal analysis of ectoderm and mesenchyme e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>FB00000902</td>\n",
       "      <td>2.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>Transcriptome Atlases of the Craniofacial Sutures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>708</th>\n",
       "      <td>FB00000903</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Transcriptome Atlases of the Craniofacial Sutures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>FB00000904</td>\n",
       "      <td>2.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>Transcriptome Atlases of the Craniofacial Sutures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>710</th>\n",
       "      <td>FB00000892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3D White Light Photogrammetry Images of North ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>711</th>\n",
       "      <td>FB00000891</td>\n",
       "      <td>2.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>Temporal analysis of ectoderm and mesenchyme e...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        accession  species  stage  \\\n",
       "700    FB00000886      NaN    NaN   \n",
       "701    FB00000887      NaN    NaN   \n",
       "702    FB00000888      NaN    NaN   \n",
       "703    FB00000889      NaN    NaN   \n",
       "704    FB00000890      NaN    NaN   \n",
       "705  FB00000807.2      3.0  109.0   \n",
       "706    FB00000867      NaN    NaN   \n",
       "707    FB00000902      2.0   25.0   \n",
       "708    FB00000903      2.0    4.0   \n",
       "709    FB00000904      2.0   25.0   \n",
       "710    FB00000892      NaN    NaN   \n",
       "711    FB00000891      2.0   30.0   \n",
       "\n",
       "                                                 title  \n",
       "700     Skull vault formation in individual WT AB Fish  \n",
       "701     Skull vault formation in individual WT AB Fish  \n",
       "702     Skull vault formation in individual WT AB Fish  \n",
       "703     Skull vault formation in individual WT AB Fish  \n",
       "704     Skull vault formation in individual WT AB Fish  \n",
       "705  ChIP-seq of multiple histone marks and RNA-seq...  \n",
       "706  Temporal analysis of ectoderm and mesenchyme e...  \n",
       "707  Transcriptome Atlases of the Craniofacial Sutures  \n",
       "708  Transcriptome Atlases of the Craniofacial Sutures  \n",
       "709  Transcriptome Atlases of the Craniofacial Sutures  \n",
       "710  3D White Light Photogrammetry Images of North ...  \n",
       "711  Temporal analysis of ectoderm and mesenchyme e...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities.dataframe[700:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Faceting Examples\n",
    "You may have noticed that in the examples above, the 'species' and 'stage' attributes are numeric. These are internal \"primary keys\" that are not meaningful and difficult to filter on. We may want to construct filters on our datasets based on these categories. This can be used for \"faceted search\" modes and can be useful even within the context of programmatic access to data in the catalog."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example: faceting on \"related\" tables\n",
    "Let's say we want to find all of the samples in our catalog where their species are 'Mus musculus' and their age stage are 'E10.5'.\n",
    "\n",
    "We need to extend our understanding of the data model with the following tables that are related to '`sample`'.\n",
    "- `isa.sample.species -> vocabulary.species.id`: the sample table has a foreign key reference to the '`species`' table.\n",
    "- `isa.sample.stage -> vocabulary.stage.id`: the sample table has a foreign key reference to the '`stage`' table.\n",
    "\n",
    "We may say that `species` and `stage` are _related_ to the `sample` table in the sense that `sample` has a direct foreign key relationship from it to them.\n",
    "\n",
    "For convenience, we will get local variables for the species and stage tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "species = pb.vocabulary.species\n",
    "stage = pb.vocabulary.stage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, let's link samples with species and filter on the term \"Mus musculus\" (i.e., \"mouse\")."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/S:=isa:sample/species:=vocabulary:species/term=Mus%20musculus\n"
     ]
    }
   ],
   "source": [
    "path = sample.alias('S').link(species).filter(species.term == 'Mus musculus')\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the _context_ of the path is the `species` table instance, but we need to link from the `sample` to the age `stage` table.\n",
    "\n",
    "To do so, we reference the `sample` table instance, in this case using its alias `S`. Then we link off of that table instance which updates the `path` itself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/entity/S:=isa:sample/species:=vocabulary:species/term=Mus%20musculus/$S/stage:=vocabulary:stage/term=E10.5\n"
     ]
    }
   ],
   "source": [
    "path.S.link(stage).filter(stage.term == 'E10.5')\n",
    "print(path.uri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, the path _context_ is the age `stage` table instance, but we wanted to get the entities for the `sample` table. To do so, again we will reference the `sample` table instance by the alias `S` we used. From there, we will call the `entities(...)` method to get the samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.facebase.org/ermrest/catalog/1/attribute/S:=isa:sample/species:=vocabulary:species/term=Mus%20musculus/$S/stage:=vocabulary:stage/term=E10.5/$S/S:id,S:collection_date,species:=species:term,species_iri:=species:iri,stage:=stage:term,stage_iei:=stage:iri\n"
     ]
    }
   ],
   "source": [
    "entities = path.S.entities(path.S.id,\n",
    "                           path.S.collection_date,\n",
    "                           species=path.species.term,\n",
    "                           species_iri=path.species.iri,\n",
    "                           stage=path.stage.term,\n",
    "                           stage_iei=path.stage.iri)\n",
    "print(entities.uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>collection_date</th>\n",
       "      <th>id</th>\n",
       "      <th>species</th>\n",
       "      <th>species_iri</th>\n",
       "      <th>stage</th>\n",
       "      <th>stage_iei</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2015-07-14</td>\n",
       "      <td>1331</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>http://purl.bioontology.org/ontology/NCBITAXON...</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>http://purl.org/sig/ont/cmo/cmo0008759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2015-07-14</td>\n",
       "      <td>1332</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>http://purl.bioontology.org/ontology/NCBITAXON...</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>http://purl.org/sig/ont/cmo/cmo0008759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2015-07-14</td>\n",
       "      <td>1333</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>http://purl.bioontology.org/ontology/NCBITAXON...</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>http://purl.org/sig/ont/cmo/cmo0008759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2015-07-14</td>\n",
       "      <td>1334</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>http://purl.bioontology.org/ontology/NCBITAXON...</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>http://purl.org/sig/ont/cmo/cmo0008759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2015-07-14</td>\n",
       "      <td>1335</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>http://purl.bioontology.org/ontology/NCBITAXON...</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>http://purl.org/sig/ont/cmo/cmo0008759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2015-07-14</td>\n",
       "      <td>1336</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>http://purl.bioontology.org/ontology/NCBITAXON...</td>\n",
       "      <td>E10.5</td>\n",
       "      <td>http://purl.org/sig/ont/cmo/cmo0008759</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  collection_date    id       species  \\\n",
       "0      2015-07-14  1331  Mus musculus   \n",
       "1      2015-07-14  1332  Mus musculus   \n",
       "2      2015-07-14  1333  Mus musculus   \n",
       "3      2015-07-14  1334  Mus musculus   \n",
       "4      2015-07-14  1335  Mus musculus   \n",
       "5      2015-07-14  1336  Mus musculus   \n",
       "\n",
       "                                         species_iri  stage  \\\n",
       "0  http://purl.bioontology.org/ontology/NCBITAXON...  E10.5   \n",
       "1  http://purl.bioontology.org/ontology/NCBITAXON...  E10.5   \n",
       "2  http://purl.bioontology.org/ontology/NCBITAXON...  E10.5   \n",
       "3  http://purl.bioontology.org/ontology/NCBITAXON...  E10.5   \n",
       "4  http://purl.bioontology.org/ontology/NCBITAXON...  E10.5   \n",
       "5  http://purl.bioontology.org/ontology/NCBITAXON...  E10.5   \n",
       "\n",
       "                                stage_iei  \n",
       "0  http://purl.org/sig/ont/cmo/cmo0008759  \n",
       "1  http://purl.org/sig/ont/cmo/cmo0008759  \n",
       "2  http://purl.org/sig/ont/cmo/cmo0008759  \n",
       "3  http://purl.org/sig/ont/cmo/cmo0008759  \n",
       "4  http://purl.org/sig/ont/cmo/cmo0008759  \n",
       "5  http://purl.org/sig/ont/cmo/cmo0008759  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entities.dataframe"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
